pkg/storage/storagewrappers/sharediterator/shared_iterator_datastore.go

```diff
@@ -296,36 +363,175 @@ func (sf *IteratorDatastore) deref(cacheKey string) (bool, error) {
+func (s *sharedIterator) watchdogTimeout() {
+	s.mu.Lock()
+	s.watchdogTimeoutErr = errSharedIteratorWatchdog
+	s.mu.Unlock()
+	s.Stop()
 }
 
 func (s *sharedIterator) Head(ctx context.Context) (*openfgav1.Tuple, error) {
-	return s.inner.Head(ctx)
+	ctx, span := tracer.Start(
+		ctx,
+		"sharedIterator.Head",
+	)
+	defer span.End()
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.watchdogTimer != nil {
+		s.watchdogTimer.Reset(s.maxAliveTime)
+	}
+
+	if s.watchdogTimeoutErr != nil {
+		return nil, s.watchdogTimeoutErr
+	}
+
+	if s.stopped.Load() {
+		span.SetAttributes(attribute.Bool("stopped", true))
+		return nil, storage.ErrIteratorDone
+	}
+
+	if s.head == len(*s.items) {
+		// If there is an error, no need to get any more items, and we just return the error
+		if s.sharedErr != nil && *s.sharedErr != nil {
+			span.SetAttributes(attribute.String("existingError", (*s.sharedErr).Error()))
+			return nil, *s.sharedErr
+		}
+
+		span.SetAttributes(attribute.Bool("newItem", true))
+		item, err := s.inner.Next(ctx)
+		if err != nil {
+			span.SetAttributes(attribute.String("newError", err.Error()))
+			*s.sharedErr = err
+			return nil, err
+		}
+		*s.items = append(*s.items, item)
+		return item, nil
+	}
+
+	span.SetAttributes(attribute.Bool("newItem", false))
+	return (*s.items)[s.head], nil
 }
 
 func (s *sharedIterator) Next(ctx context.Context) (*openfgav1.Tuple, error) {
-	return s.inner.Next(ctx)
+	ctx, span := tracer.Start(
+		ctx,
+		"sharedIterator.Next",
+	)
+	defer span.End()
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+
+	if s.watchdogTimer != nil {
+		s.watchdogTimer.Reset(s.maxAliveTime)
+	}
+
+	if s.watchdogTimeoutErr != nil {
+		return nil, s.watchdogTimeoutErr
+	}
+
+	if s.stopped.Load() {
+		span.SetAttributes(attribute.Bool("stopped", true))
+		return nil, storage.ErrIteratorDone
+	}
+	if s.head < len(*s.items) {
+		currentHead := s.head
+		s.head++
+		return (*s.items)[currentHead], nil
+	}
+	if s.sharedErr != nil && *s.sharedErr != nil {
+		// we are going to get more items. However, if we see error
+		// previously, we will not need to get more items
+		// and can simply return the same error.
+		span.SetAttributes(attribute.String("existingError", (*s.sharedErr).Error()))
+
+		return nil, *s.sharedErr
+	}
+	span.SetAttributes(attribute.Bool("newItem", true))
+
+	item, err := s.inner.Next(ctx)
+	if err != nil {
+		span.SetAttributes(attribute.String("newError", err.Error()))
+		*s.sharedErr = err
+		return nil, err
+	}
+	*s.items = append(*s.items, item)
+	s.head++
+	return item, nil
 }
 
 func (s *sharedIterator) Stop() {
-	// For now, we don't check whether this is the last item in the shared iterator
-	// as this is just a skeleton.  When the actual shared iterator is implemented,
-	// we will stop the inner iterator only if this is the last reference.
-	if !s.stopped.Swap(true) {
+	s.mu.Lock()
+	if s.watchdogTimer != nil {
+		s.watchdogTimer.Stop()
+	}
+	s.mu.Unlock()
+
+	if s.stopped.Swap(true) {
```

Comment:
If you already have a `mu` I don't particularly see why this has to be an atomic 🤷🏽

Reply:
yes, with mu, we can change this to non-atomic.